### PR TITLE
feat(destinations): add GCP destinations

### PR DESCRIPTION
**Pull request by gjanco**
Adding GCP Cloud Monitoring, Cloud
Operations and Cloud PubSub Sinks

Ref: LOG-20340

### CODE REVIEW

**Comment by darinspivey** [comment]
@gjanco, nbd, but please make all git branches include your username and card, `gjanco/LOG-20340` (unless it's some sort of hotfix without a card)

**Review by darinspivey — COMMENTED** [review]
Looks good, mostly nits.  All of my comments pretty much refer to all 3 destinations, so if you choose to make any of those changes, please apply to all of them.

File: examples/resources/mezmo_gcp_cloud_monitoring_destination/resource.tf [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+resource "mezmo_gcp_cloud_monitoring_destination" "gcp" {
+  title            = "GCP Cloud Monitoring"
+  description      = "This stores our metrics events in GCP cloud monitoring"
+  inputs           = [mezmo_demo_source.source1.id]
+  pipeline_id      = mezmo_pipeline.pipeline1.id
+  resource_type    = "global2"
+  project_id       = "proj456"
+  credentials_json = "{}"
```
Comment (biblicalph) [inline examples/resources/mezmo_gcp_cloud_monitoring_destination/resource.tf:32]: Let's use an actual value here

File: examples/resources/mezmo_gcp_cloud_operations_destination/resource.tf [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+resource "mezmo_gcp_cloud_operations_destination" "gcp" {
+  title            = "GCP Cloud Operations"
+  description      = "This stores our log events in GCP cloud operations"
+  inputs           = [mezmo_demo_source.source1.id]
+  pipeline_id      = mezmo_pipeline.pipeline1.id
+  resource_type    = "global2"
+  project_id       = "proj456"
+  log_id           = "log456"
+  credentials_json = "{}"
```
Comment (biblicalph) [inline examples/resources/mezmo_gcp_cloud_operations_destination/resource.tf:33]: And here

File: examples/resources/mezmo_gcp_cloud_pubsub_destination/resource.tf [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+resource "mezmo_gcp_cloud_pubsub_destination" "gcp" {
+  title            = "GCP Cloud PubSub"
+  description      = "This stores our log events in GCP cloud PubSub"
+  inputs           = [mezmo_demo_source.source1.id]
+  pipeline_id      = mezmo_pipeline.pipeline1.id
+  encoding         = "text"
+  project_id       = "proj456"
+  topic            = "topic456"
+  credentials_json = "{}"
```
Comment (biblicalph) [inline examples/resources/mezmo_gcp_cloud_pubsub_destination/resource.tf:33]: And here

File: internal/provider/models/destinations/gcp_cloud_monitoring.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		if len(labelsArray) > 0 {
+			labelMap := make(map[string]any, len(labelsArray))
+			for _, obj := range labelsArray {
+				obj := obj.(map[string]any)
+				key := obj["label_name"].(string)
+				value := obj["label_value"].(string)
+				labelMap[key] = value
+			}
+			labelType := plan.ResourceLabels.ElementType(context.Background())
+			if labelType == nil {
```
Comment (biblicalph) [inline internal/provider/models/destinations/gcp_cloud_monitoring.go:135]: Wondering if we should skip this check and always get the type from the schema  given that plan.ResourceLabels may not be set

File: internal/provider/models/destinations/gcp_cloud_operations.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+		if len(labelsArray) > 0 {
+			labelMap := make(map[string]any, len(labelsArray))
+			for _, obj := range labelsArray {
+				obj := obj.(map[string]any)
+				key := obj["label_name"].(string)
+				value := obj["label_value"].(string)
+				labelMap[key] = value
+			}
+			labelType := plan.ResourceLabels.ElementType(context.Background())
+			if labelType == nil {
```
Comment (biblicalph) [inline internal/provider/models/destinations/gcp_cloud_operations.go:145]: same here

**Review by biblicalph — COMMENTED** [review]
Wondering if pipeline service should be rejecting credentials json specified as `{}`

File: docs/resources/gcp_cloud_monitoring_destination.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+resource "mezmo_gcp_cloud_monitoring_destination" "gcp" {
+  title            = "GCP Cloud Monitoring"
+  description      = "This stores our metrics events in GCP cloud monitoring"
+  inputs           = [mezmo_demo_source.source1.id]
+  pipeline_id      = mezmo_pipeline.pipeline1.id
+  resource_type    = "global2"
+  project_id       = "proj456"
+  credentials_json = "{}"
```
[the remark this answers is not in the document]
Comment (gjanco) [reply docs/resources/gcp_cloud_monitoring_destination.md:47]: This is not a pipeline service credential.  This is a GCP credential.  And, yes, I can copy in my local test one and just leave out the private key.

File: internal/provider/models/destinations/gcp_cloud_monitoring.go (hunk 4 above)
Comment (gjanco) [reply internal/provider/models/destinations/gcp_cloud_monitoring.go:135]: Going with this suggestion, @darinspivey - just to be safe (still getting rid of the conditional).

Comment (darinspivey) [reply internal/provider/models/destinations/gcp_cloud_monitoring.go:135]: Yep, that's fine

File: internal/provider/models/destinations/gcp_cloud_monitoring.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		labelsArray, _ := component.UserConfig["resource_labels"].([]any)
+		if len(labelsArray) > 0 {
+			labelMap := make(map[string]any, len(labelsArray))
+			for _, obj := range labelsArray {
+				obj := obj.(map[string]any)
+				key := obj["label_name"].(string)
+				value := obj["label_value"].(string)
+				labelMap[key] = value
+			}
+			labelType := GcpCloudMonitoringResourceSchema.Attributes["resource_labels"].GetType().(basetypes.MapType).ElemType
```
Comment (darinspivey) [inline internal/provider/models/destinations/gcp_cloud_monitoring.go:132]: There's a duplicate `basetypes` import up top.  Use the `.` one by changing this to `MapType` and remove the other import.

File: internal/provider/models/destinations/gcp_cloud_monitoring.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+			plan.ResourceLabels = basetypes.NewMapValueMust(labelType, MapAnyToMapValues(labelMap))
```
Comment (darinspivey) [inline internal/provider/models/destinations/gcp_cloud_monitoring.go:133]: Ditto.  Remove anything that's `basetypes.xxx`

File: internal/provider/models/destinations/gcp_cloud_operations.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
+package destinations
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework-validators/mapvalidator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	. "github.com/hashicorp/terraform-plugin-framework/types/basetypes"
```
Comment (darinspivey) [inline internal/provider/models/destinations/gcp_cloud_operations.go:10]: Same duplicate import.  Remove references to `basetype.`

File: internal/provider/models/destinations/test/gcp_cloud_monitoring_test.go [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+						"mezmo_gcp_cloud_monitoring_destination.my_dest", "id", regexp.MustCompile(`[\w-]{36}`)),
+					StateHasExpectedValues("mezmo_gcp_cloud_monitoring_destination.my_dest", map[string]any{
+						"pipeline_id":       "#mezmo_pipeline.test_parent.id",
+						"title":             "new test dest",
+						"description":       "this is a new test description",
+						"generation_id":     "1",
+						"ack_enabled":       "true",
+						"inputs.#":          "1",
+						"resource_type":     "global2",
+						"resource_labels.%": nil,
```
Comment (darinspivey) [inline internal/provider/models/destinations/test/gcp_cloud_monitoring_test.go:148]: This should be fine, but FYI, after I wrote this I needed a similar check.  `TestCheckNoResourceAttr` is designed for this.  If you were to use it for this property, it'd be its own assertion:
```
resource.TestCheckNoResourceAttr("mezmo_gcp_cloud_monitoring_destination.my_dest", "resource_labels")
```

**Review by darinspivey — COMMENTED** [review]
LGTM, just a nit on imports

**Review by darinspivey — APPROVED** [review]
Thanks!

**Pull request merged**